File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.java
class Solution {
    public long[] resultArray(int[] nums, int k) {
        long[] result = new long[k];
        // dp[r] represents the count of subarrays ending at the current index with product % k == r
        long[] dp = new long[k];

        for (int num : nums) {
            long[] nextDp = new long[k];
            int modVal = num % k;

            // Single-element subarray [num]
            nextDp[modVal]++;

            // Extend previous subarrays
            for (int r = 0; r < k; r++) {
                if (dp[r] > 0) {
                    int newRem = (int) (((long) r * modVal) % k);
                    nextDp[newRem] += dp[r];
                }
            }

            // Accumulate to total result
            for (int r = 0; r < k; r++) {
                result[r] += nextDp[r];
            }

            dp = nextDp;
        }

        return result;
    }
}

File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.py
class Solution:
    def resultArray(self, nums: list[int], k: int) -> list[int]:
        result = [0] * k
        dp = [0] * k  # dp[r] is the count of subarrays ending at current index with product % k == r

        for num in nums:
            next_dp = [0] * k
            mod_val = num % k

            # Subarray consisting of just [num]
            next_dp[mod_val] += 1

            # Extend previous subarrays
            for r in range(k):
                if dp[r]:
                    new_rem = (r * mod_val) % k
                    next_dp[new_rem] += dp[r]

            # Accumulate into overall result
            for r in range(k):
                result[r] += next_dp[r]

            dp = next_dp

        return result